### PR TITLE
Implement all Hash#each* methods

### PR DESCRIPTION
**Pull request by richardboehme**
Implements:
* `Hash#each_key`
* `Hash#each_value`

Add spec for `Hash#each_pair`

### CODE REVIEW

File: src/hash_value.cpp [hunk 1]
```diff
@@ -378,6 +378,22 @@ ValuePtr HashValue::each(Env *env, Block *block) {
     return this;
 }
 
+ValuePtr HashValue::each_key(Env *env, Block *block) {
+    if (!block)
+        return send(env, SymbolValue::intern("enum_for"), { SymbolValue::intern("each_key") });
+
+    keys(env)->send(env, SymbolValue::intern("each"), {}, block);
+    return this;
+}
```
Comment (seven1m) [inline src/hash_value.cpp:387]: What do you think about implementing this in `hash.rb` so we can avoid the creation of the array of keys:

```ruby
def each_key
  each do |key, _|
    yield key
  end
end
```

File: src/hash_value.cpp [hunk 2]
```diff
@@ -378,6 +378,22 @@ ValuePtr HashValue::each(Env *env, Block *block) {
...
+
+ValuePtr HashValue::each_value(Env *env, Block *block) {
+    if (!block)
+        return send(env, SymbolValue::intern("enum_for"), { SymbolValue::intern("each_value") });
+
+    values(env)->send(env, SymbolValue::intern("each"), {}, block);
+    return this;
+}
```
Comment (seven1m) [inline src/hash_value.cpp:395]: Same:

```ruby
def each_value
  each do |_, value|
    yield value
  end
end
```

**Review by seven1m — CHANGES_REQUESTED** [review]
Just some thoughts I had about avoiding the creation of the array in both cases.

I'm thinking it would be more efficient when the user program breaks out of the loop:

```ruby
h = {...} # some huge hash
h.each_key do |key|
  # some work...
  break if key == :foo
end
```

File: src/hash_value.cpp (hunk 1 above)
Comment (seven1m) [reply src/hash_value.cpp:387]: I should say, you can obviously do that in C++ too, but it was just easier for me to illustrate in Ruby quickly. :wink: Really whichever language you want to implement it in...

**Comment by richardboehme** [comment]
Yeah you're absolutely right. I'm going to change that!

**Review by seven1m — APPROVED** [review]

**Pull request merged**